### PR TITLE
objectstore: move vector::resever in collection_list.

### PR DESCRIPTION
**Pull request by majianpeng**

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -3671,7 +3671,6 @@ void OSD::recursive_remove_collection(CephContext* cct,
   ghobject_t next;
   int max = cct->_conf->osd_target_transaction_size;
   vector<ghobject_t> objects;
-  objects.reserve(max);
```
Comment (tchaikov) [inline src/osd/OSD.cc]: take `BlueStore::collection_list()` as an example, by moving `objects.reserve(max)` into `BlueStore::_collection_list()`,  `objects.reserve(max)` will be called multiple times here. so what's the merit of this change?

**Comment by majianpeng** [comment]
About the failed read for make check.  In test it set max=INT_MAX(tcmalloc: large alloc 309237645312 bytes == (nil) ).  So failed .

**Comment by yuriw** [comment]
@majianpeng @neha-ojha needs rebase

--- pr 21848 --- pulling https://github.com/majianpeng/ceph.git branch objectstore-collection-list
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Total 19 (delta 15), reused 15 (delta 15), pack-reused 4
Unpacking objects: 100% (19/19), done.
From https://github.com/majianpeng/ceph
 * branch            objectstore-collection-list -> FETCH_HEAD
Auto-merging src/osd/PrimaryLogPG.cc
Auto-merging src/osd/OSD.cc
Auto-merging src/os/memstore/MemStore.cc
Auto-merging src/os/kstore/KStore.cc
Auto-merging src/os/filestore/HashIndex.cc
Auto-merging src/os/bluestore/BlueStore.cc
CONFLICT (file/directory): There is a directory with name src/dmclock in 7cc3b450f1560cb8b979fe4f895d07a2a5382834. Adding src/dmclock as src/dmclock~HEAD
Adding qa/suites/multimds/thrash/mount
Adding qa/suites/multimds/basic/mount
Automatic merge failed; fix conflicts and then commit the result.
Traceback (most recent call last):
  File "/home/yuriw/wip_master/src/script/build-integration-branch", line 62, in <module>
    assert not r
AssertionError

**Comment by neha-ojha** [comment]
retest this please

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-10-30-1642](https://trello.com/c/rtcu9E3q/455-wip-yuri3-testing-2018-10-30-1642)

**Comment by yuriw** [comment]
@majianpeng @neha-ojha @tchaikov @jdurgin 
I suspect this PR broke the build, see 

/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.0-4692-ge717dfc/rpm/el7/BUILD/ceph-14.0.0-4692-ge717dfc/src/test/objectstore/store_test.cc: In member function 'virtual void StoreTest_ManyObjectTest_Test::TestBody()':
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.0.0-4692-ge717dfc/rpm/el7/BUILD/ceph-14.0.0-4692-ge717dfc/src/test/objectstore/store_test.cc:3283:16: error: overloaded function with no contextual type information
   while (!next.is_max()) {
                ^~~~~~

in https://shaman.ceph.com/builds/ceph/wip-yuri3-testing-2018-10-30-1642/

**Comment by neha-ojha** [comment]
the `make check` failure https://jenkins.ceph.com/job/ceph-pull-requests/8096/consoleFull#4909035072a811ea2-3e7b-466b-84b4-d13df7e35809 looks related

**Comment by tchaikov** [comment]
retest this please

**Comment by majianpeng** [comment]
@tchaikov。 failed cause code. Now fixing.  sorry for that.

**Comment by majianpeng** [comment]
update && passed make check.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8437,6 +8437,7 @@ int BlueStore::_collection_list(
     }
   }
   dout(20) << __func__ << " pend " << pretty_binary_string(pend) << dendl;
+  ls->reserve(max);
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]:
```
 ceph version 14.0.1-771-g680ef7c (680ef7c847bac83761172d65ca75edda38014eba) nautilus (dev)
 1: (()+0xf5d0) [0x7f22274d85d0]
 2: (gsignal()+0x37) [0x7f22262cf207]
 3: (abort()+0x148) [0x7f22262d08f8]
 4: (__gnu_cxx::__verbose_terminate_handler()+0x165) [0x7f2226bde7d5]
 5: (()+0x5e746) [0x7f2226bdc746]
 6: (()+0x5e773) [0x7f2226bdc773]
 7: (()+0x5e993) [0x7f2226bdc993]
 8: (()+0x1772b) [0x7f2229e2a72b]
 9: (tcmalloc::allocate_full_cpp_throw_oom(unsigned long)+0xf3) [0x7f2229e48a03]
 10: (HashIndex::list_by_hash_bitwise(std::vector<std::string, std::allocator<std::string> > const&, ghobject_t const&, int, ghobject_t*, std::
vector<ghobject_t, std::allocator<ghobject_t> >*)+0x853) [0x55af0d626f53]
 11: (HashIndex::_collection_list_partial(ghobject_t const&, ghobject_t const&, int, std::vector<ghobject_t, std::allocator<ghobject_t> >*, gho
bject_t*)+0x12b) [0x55af0d62753b]
 12: (FileStore::collection_list(coll_t const&, ghobject_t const&, ghobject_t const&, int, std::vector<ghobject_t, std::allocator<ghobject_t> >
*, ghobject_t*)+0x259) [0x55af0d443c59]
 13: (PGBackend::objects_list_range(hobject_t const&, hobject_t const&, std::vector<hobject_t, std::allocator<hobject_t> >*, std::vector<ghobje
ct_t, std::allocator<ghobject_t> >*)+0x147) [0x55af0d2d50e7]
 14: (PG::build_scrub_map_chunk(ScrubMap&, ScrubMapBuilder&, hobject_t, hobject_t, bool, ThreadPool::TPHandle&)+0x28a) [0x55af0d18812a]
 15: (PG::chunky_scrub(ThreadPool::TPHandle&)+0x15bd) [0x55af0d1b035d]
 16: (PG::scrub(unsigned int, ThreadPool::TPHandle&)+0xaf) [0x55af0d1b143f]
 17: (PGScrub::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x12) [0x55af0d35d6b2]
 18: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xb6c) [0x55af0d0e9dbc]
```

this breaks tcmalloc. 

see

https://github.com/ceph/ceph/blob/0aa5566c81b09175971644f9d85e665919b8abef/src/osd/PGBackend.cc#L388-L394

we cannot `reserve(max)` . sometimes `max` is really large.

Comment (majianpeng) [reply src/os/bluestore/BlueStore.cc]: i forget this INT_MAX.  i modified some places which used INT_MAX in tests. I think using INT_MAX is  a bug which used wrong parameters. If  collection has many objects, it still allocate memory.
ObjectStore provois this data:https://github.com/ceph/ceph/blob/0aa5566c81b09175971644f9d85e665919b8abef/src/os/ObjectStore.h#L1597.  So we can use this.

**Comment by neha-ojha** [comment]
@majianpeng looks like we should retest this in teuthology, can you please rebase it?

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
seems it does not compile.

retest this please.

**Comment by majianpeng** [comment]
@tchaikov . i'll fix.

**Comment by majianpeng** [comment]
@tchaikov  .fixed .

**Comment by liewegas** [comment]
suspicious failure:
```
"2019-04-23 18:49:20.368657 osd.3 (osd.3) 544 : cluster [ERR] 6.1s1 shard 6(0) 6:80520673:::benchmark_data_smithi091_206143_object95842:head : missing" in cluster log
```
/a/sage-2019-04-23_06:50:56-rados-wip-sage4-testing-2019-04-22-2147-distro-basic-smithi/3882552

**Comment by liewegas** [comment]
actually,
```
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80520673:::benchmark_data_smithi091_206143_object95842:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:8082360d:::benchmark_data_smithi091_206143_object93501:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:8082bb7a:::benchmark_data_smithi091_206143_object100512:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80838a38:::benchmark_data_smithi091_206143_object98123:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80854d1d:::benchmark_data_smithi091_206143_object100209:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:808c440e:::benchmark_data_smithi091_206143_object94299:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80a0fdb8:::benchmark_data_smithi091_206143_object98641:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80a75a36:::benchmark_data_smithi091_206143_object98388:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80b46788:::benchmark_data_smithi091_206143_object96017:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80b58578:::benchmark_data_smithi091_206143_object99671:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80c57304:::benchmark_data_smithi091_206143_object94400:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80d19b04:::benchmark_data_smithi091_206143_object96941:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80d33acb:::benchmark_data_smithi091_206143_object93577:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80d4074a:::benchmark_data_smithi091_206143_object98193:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80d85a6f:::benchmark_data_smithi091_206143_object100729:head : missing
2019-04-23 18:49:20.367 7f4a98351700 -1 log_channel(cluster) log [ERR] : 6.1s1 shard 6(0) 6:80db070f:::benchmark_data_smithi091_206143_object97653:head : missing
```

**Pull request closed without merging**